src/MongoDB.Driver/TransactionExecutor.cs

```diff
@@ -0,0 +1,245 @@
+                if (transactionHasBeenCommitted)
+                {
+                    return callbackOutcome.Result;
+                }
+            }
+        }
+
+        private static bool HasTimedOut(DateTime startTime, DateTime currentTime)
+        {
+            return (currentTime - startTime) >= __transactionTimeout;
+        }
+
+        private static CallbackOutcome<TResult> ExecuteCallback<TResult>(IClientSessionHandle clientSession, Func<IClientSessionHandle, CancellationToken, TResult> callback, DateTime startTime, IClock clock, CancellationToken cancellationToken)
+        {
+            try
+            {
+                var result = callback(clientSession, cancellationToken);
+                return new CallbackOutcome<TResult>.WithResult(result);
+            }
+            catch (Exception ex)
+            {
+                if (IsTransactionInStartingOrInProgressState(clientSession))
+                {
+                    clientSession.AbortTransaction(cancellationToken);
+                }
+
+                if (HasErrorLabel(ex, TransientTransactionErrorLabel) && !HasTimedOut(startTime, clock.UtcNow))
+                {
+                    return new CallbackOutcome<TResult>.WithShouldRetryTransaction();
+                }
+
+                throw;
+            }
+        }
+
+        private static async Task<CallbackOutcome<TResult>> ExecuteCallbackAsync<TResult>(IClientSessionHandle clientSession, Func<IClientSessionHandle, CancellationToken, Task<TResult>> callbackAsync, DateTime startTime, IClock clock, CancellationToken cancellationToken)
+        {
+            try
+            {
+                var result = await callbackAsync(clientSession, cancellationToken).ConfigureAwait(false);
+                return new CallbackOutcome<TResult>.WithResult(result);
+            }
+            catch (Exception ex)
+            {
+                if (IsTransactionInStartingOrInProgressState(clientSession))
+                {
+                    await clientSession.AbortTransactionAsync(cancellationToken).ConfigureAwait(false);
+                }
+
+                if (HasErrorLabel(ex, TransientTransactionErrorLabel) && !HasTimedOut(startTime, clock.UtcNow))
+                {
+                    return new CallbackOutcome<TResult>.WithShouldRetryTransaction();
+                }
+
+                throw;
+            }
+        }
+
+        private static bool CommitWithRetries(IClientSessionHandle clientSession, DateTime startTime, IClock clock, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                try
+                {
+                    clientSession.CommitTransaction(cancellationToken);
+                    return true;
+                }
+                catch (Exception ex)
+                {
+                    var now = clock.UtcNow;
```

Comment:
It would be worth adding a comment:

    var now = clock.UtcNow; // call UtcNow once whether we need to or not to facilitate predictable mocking

Reply:
Done